### PR TITLE
Always lazy load comment meta.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/57801

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/comment.php [hunk 1]
```diff
@@ -3345,7 +3357,11 @@ function _prime_comment_caches( $comment_ids, $update_meta_cache = true ) {
 	if ( ! empty( $non_cached_ids ) ) {
 		$fresh_comments = $wpdb->get_results( sprintf( "SELECT $wpdb->comments.* FROM $wpdb->comments WHERE comment_ID IN (%s)", implode( ',', array_map( 'intval', $non_cached_ids ) ) ) );
 
-		update_comment_cache( $fresh_comments, $update_meta_cache );
+		update_comment_cache( $fresh_comments, false );
+	}
+
+	if ( $update_meta_cache ) {
```
Comment (mukeshpanchal27) [inline src/wp-includes/comment.php:3367]: Similar to `_prime_term_caches` function we have to add `@since 6.3.0 Use wp_lazyload_comment_meta() for lazy-loading of term meta.` document.

File: src/wp-includes/comment-template.php [hunk 2]
```diff
@@ -1440,12 +1440,11 @@ function comments_template( $file = '/comments.php', $separate_comments = false
 	$comment_author_url = esc_url( $commenter['comment_author_url'] );
 
 	$comment_args = array(
-		'orderby'                   => 'comment_date_gmt',
-		'order'                     => 'ASC',
-		'status'                    => 'approve',
-		'post_id'                   => $post->ID,
-		'no_found_rows'             => false,
-		'update_comment_meta_cache' => false, // We lazy-load comment meta for performance.
```
Comment (mukeshpanchal27) [inline src/wp-includes/comment-template.php:1448]: Q: Do we needs to remove `update_comment_meta_cache` from `get_comment_count()` and `wp_comments_personal_data_exporter()` function?

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @spacedmonkey, The changes look solid to me. Left one minor feedback and question.

**Before:**
2.27s  16.9MB  0.06s  30Q

**After:**
2.19s  16.9MB  0.05s  29Q

File: src/wp-includes/comment-template.php (hunk 2 above)
Comment (spacedmonkey) [reply src/wp-includes/comment-template.php:1448]: @mukeshpanchal27 No, `get_comment_count` and `wp_comments_personal_data_exporter` are set to false, as comment meta is not need. This is set to false as `wp_queue_comments_for_comment_meta_lazyload` was called. `wp_queue_comments_for_comment_meta_lazyload` is no longer needed.

File: src/wp-includes/comment.php [hunk 3]
```diff
@@ -484,6 +484,21 @@ function get_comment_meta( $comment_id, $key = '', $single = false ) {
...
 }
 
+/**
+ * Queue comment meta for lazy-loading.
+ *
+ * @since 6.3.0
+ *
+ * @param array $comment_ids List of comment IDs.
+ */
+function wp_lazyload_comment_meta( array $comment_ids ) {
```
Comment (peterwilsoncc) [inline src/wp-includes/comment.php:494]: This seems very similar to `wp_queue_comments_for_comment_meta_lazyload()`, why is it required?

File: tests/phpunit/tests/comment/metaCache.php [hunk 4]
```diff
@@ -12,8 +12,6 @@ class Tests_Comment_MetaCache extends WP_UnitTestCase {
 	 * @covers ::update_comment_meta
 	 */
 	public function test_update_comment_meta_cache_should_default_to_true() {
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/comment/metaCache.php]:
```suggestion
	public function test_update_comment_meta_cache_should_default_to_lazy_loading() {
```

File: tests/phpunit/tests/comment/metaCache.php [hunk 5]
```diff
@@ -24,18 +22,21 @@ public function test_update_comment_meta_cache_should_default_to_true() {
...
-		$q = new WP_Comment_Query(
+		$num_queries = get_num_queries();
+		$q           = new WP_Comment_Query(
 			array(
 				'post_ID' => $p,
 			)
 		);
 
-		$num_queries = $wpdb->num_queries;
+		$this->assertSame( $num_queries + 2, get_num_queries() );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/comment/metaCache.php]:
```suggestion
		$this->assertSame( 2, get_num_queries() - $num_queries, 'Querying comments is expected to make two queries' );
```

File: tests/phpunit/tests/comment/metaCache.php [hunk 6]
```diff
@@ -24,18 +22,21 @@ public function test_update_comment_meta_cache_should_default_to_true() {
...
+
+		$num_queries = get_num_queries();
 		foreach ( $comment_ids as $cid ) {
 			get_comment_meta( $cid, 'foo', 'bar' );
 		}
 
-		$this->assertSame( $num_queries, $wpdb->num_queries );
+		$this->assertSame( $num_queries + 1, get_num_queries() );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/comment/metaCache.php]:
```suggestion
		$this->assertSame( 1, get_num_queries() - $num_queries, 'Comment meta is expected to be lazy loaded' );
```

File: tests/phpunit/tests/comment/metaCache.php [hunk 7]
```diff
@@ -56,19 +55,59 @@ public function test_update_comment_meta_cache_true() {
...
+		$this->assertSame( $num_queries + 2, get_num_queries() );
 
-		$num_queries = $wpdb->num_queries;
+		$num_queries = get_num_queries();
 		foreach ( $comment_ids as $cid ) {
 			get_comment_meta( $cid, 'foo', 'bar' );
 		}
 
-		$this->assertSame( $num_queries, $wpdb->num_queries );
+		$this->assertSame( $num_queries + 1, get_num_queries() );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/comment/metaCache.php]: As above, this test change will need:

* messages for the multiple assertions
* assert the difference in comment count (ie 2 and 1 respectively) as it make the test outcome clearer.

File: tests/phpunit/tests/comment/metaCache.php [hunk 8]
```diff
@@ -56,19 +55,59 @@ public function test_update_comment_meta_cache_true() {
...
+			);
+			$this->assertSame( $num_queries + 1, get_num_queries(), 'Comment query should only add one query' );
+		}
+
+		$filter = new MockAction();
+		add_filter( 'update_comment_metadata_cache', array( $filter, 'filter' ), 10, 2 );
+		$num_queries = get_num_queries();
+		get_comment_meta( $comment_ids[0], 'foo', 'bar' );
+
+		$this->assertSame( $num_queries + 1, get_num_queries(), 'Comment meta should be loaded in one database query' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/comment/metaCache.php]:
```suggestion
		$this->assertSame( 1, get_num_queries() - $num_queries, 'Comment meta should be loaded in one database query' );
```

File: src/wp-includes/comment.php [hunk 9]
```diff
@@ -3345,7 +3357,11 @@ function _prime_comment_caches( $comment_ids, $update_meta_cache = true ) {
 	if ( ! empty( $non_cached_ids ) ) {
 		$fresh_comments = $wpdb->get_results( sprintf( "SELECT $wpdb->comments.* FROM $wpdb->comments WHERE comment_ID IN (%s)", implode( ',', array_map( 'intval', $non_cached_ids ) ) ) );
 
-		update_comment_cache( $fresh_comments, $update_meta_cache );
+		update_comment_cache( $fresh_comments, false );
+	}
+
+	if ( $update_meta_cache ) {
+		wp_lazyload_comment_meta( $comment_ids );
```
Comment (peterwilsoncc) [inline src/wp-includes/comment.php:3368]: Is there any reason that similar can't be done in `update_comment_cache()` too?

I'm in two minds as to whether that's a good idea or a terrible idea so would like your thoughts.

**Review by peterwilsoncc — COMMENTED** [review]
Added a few notes and a question.

File: src/wp-includes/comment.php (hunk 3 above)
Comment (spacedmonkey) [reply src/wp-includes/comment.php:494]: The function `wp_queue_comments_for_comment_meta_lazyload` is now a wrapper for `wp_lazyload_comment_meta`.  `wp_queue_comments_for_comment_meta_lazyload` expects an array of comment objects. `wp_lazyload_comment_meta` use array of ids. We also have [wp_lazyload_term_meta](https://github.com/WordPress/wordpress-develop/blob/5e4d1bc2d65fa3efa4585c6620c12a55d52d57b3/src/wp-includes/taxonomy.php#L1441) and this follows this naming structure.

File: tests/phpunit/tests/comment/metaCache.php [hunk 10]
```diff
@@ -56,19 +55,59 @@ public function test_update_comment_meta_cache_true() {
...
+			}
+
+			$num_queries = get_num_queries();
+			$q           = new WP_Comment_Query(
+				array(
+					'post_ID'                   => $p,
+					'update_comment_meta_cache' => true,
+				)
+			);
+			$this->assertSame( $num_queries + 1, get_num_queries(), 'Comment query should only add one query' );
```
Comment (spacedmonkey) [inline tests/phpunit/tests/comment/metaCache.php]:
```suggestion
			$this->assertSame( 1, get_num_queries() - $num_queries, 'Comment query should only add one query' );
```

File: src/wp-includes/comment.php (hunk 9 above)
Comment (spacedmonkey) [reply src/wp-includes/comment.php:3368]: If you look at this - https://github.com/WordPress/wordpress-develop/commit/2c6bf77a9e7532c6ac3a03502b44d108cae86179

We could do that here. If you are a developer that is `update_comment_cache`, it is likely you know what you are doing,I don't want to break that for anyone.

TLDR. Let's leave for now and maybe think about this in the future.

File: src/wp-includes/comment.php (hunk 1 above)
Comment (spacedmonkey) [reply src/wp-includes/comment.php:3367]: Fixed.

**Comment by spacedmonkey** [comment]
Awaiting review from @peterwilsoncc or @mukeshpanchal27

File: src/wp-includes/class-wp-query.php [hunk 11]
```diff
@@ -2813,7 +2813,7 @@ public function get_posts() {
 				$comment_ids = $wpdb->get_col( $comments_request );
 				wp_cache_add( $cache_key, $comment_ids, 'comment-queries' );
 			}
-			_prime_comment_caches( $comment_ids, false );
+			_prime_comment_caches( $comment_ids );
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php:2816]: Why this change? Seems unrelated?

File: src/wp-includes/class-wp-query.php [hunk 12]
```diff
@@ -3372,7 +3372,7 @@ public function get_posts() {
 				$comment_ids = $wpdb->get_col( $comments_request );
 				wp_cache_add( $comment_cache_key, $comment_ids, 'comment-queries' );
 			}
-			_prime_comment_caches( $comment_ids, false );
+			_prime_comment_caches( $comment_ids );
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php:3375]: Same question here.

File: src/wp-includes/class-wp-query.php [hunk 13]
```diff
@@ -3487,11 +3487,6 @@ public function get_posts() {
 			}
 		}
 
-		// If comments have been fetched as part of the query, make sure comment meta lazy-loading is set up.
-		if ( ! empty( $this->comments ) ) {
-			wp_queue_comments_for_comment_meta_lazyload( $this->comments );
-		}
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php:3493]: With this being removed, what replaces it? How are comments now lazy-loaded when using `WP_Query`?

File: src/wp-includes/comment.php (hunk 9 above)
Comment (felixarntz) [reply src/wp-includes/comment.php:3368]: Fair point, I agree for backward compatibility it's better to not modify that in `update_comment_cache()`, although I wish we could for a cleaner API :)

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey Left a bit of feedback, note that I'm lacking some context and therefore would love to get some clarification on a few bits.

File: src/wp-includes/class-wp-query.php (hunk 13 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php:3493]: Lazy loading is now implemented in `_prime_comment_caches`. Now that `_prime_comment_caches` is called with `update_meta_cache` true, this is no longer needed.

File: src/wp-includes/class-wp-query.php (hunk 11 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php:2816]: See https://github.com/WordPress/wordpress-develop/pull/4128/files#r1185887963

Setting this to true ( deafult value ), means comments will be lazy loaded here, meaning `wp_queue_comments_for_comment_meta_lazyload` is not needed.

File: src/wp-includes/comment-template.php [hunk 14]
```diff
@@ -2380,8 +2379,6 @@ function wp_list_comments( $args = array(), $comments = null ) {
 		$parsed_args['reverse_top_level'] = ( 'desc' === get_option( 'comment_order' ) );
 	}
 
-	wp_queue_comments_for_comment_meta_lazyload( $_comments );
```
Comment (spacedmonkey) [inline src/wp-includes/comment-template.php:2383]: This function is now not used in core. Wonder if we should a note in the phpdocs for this function.

Comment (felixarntz) [reply src/wp-includes/comment-template.php:2383]: If it's not used, maybe worth deprecating? I doubt it is seeing much use in plugins, so deprecating would seem reasonable to me. It also makes the existence of the other `wp_lazyload_comment_meta()` function less confusing.

**Review by felixarntz — COMMENTED** [review]
With the clarifications, this basically looks good to go to me, though I would like for someone else to also review this as I'm not terribly familiar with the underlying parts of the codebase.

Separately, should we deprecate `wp_queue_comments_for_comment_meta_lazyload()` if it's now unused?

File: src/wp-includes/comment-template.php (hunk 14 above)
Comment (spacedmonkey) [reply src/wp-includes/comment-template.php:2383]: @peterwilsoncc Thoughts on deprecating This function?

Comment (peterwilsoncc) [reply src/wp-includes/comment-template.php:2383]: This docblock would need to be updated as part of the process

https://github.com/WordPress/wordpress-develop/blob/4cb3e1582a022e2d59ba3c4bf8dcb5c60c45eefc/src/wp-includes/class-wp-query.php#L4883-L4896

I did a search in the [plugin repo](https://wpdirectory.net/search/01GZZD0RE5AK0CMSVFVNSM6FZ3), the [theme repo](https://wpdirectory.net/search/01GZZD3C7CZYMXWEVMF0B840BP) and [on GitHub](https://github.com/search?q=wp_queue_comments_for_comment_meta_lazyload&type=code). It doesn't appear to be used a great deal so I think it's safe to deprecate.

There's a similarly named function `wp_queue_posts_for_term_meta_lazyload()` so there's some potential for confusion as to why one remains but not the other.

File: src/wp-admin/includes/class-wp-comments-list-table.php [hunk 15]
```diff
@@ -165,8 +165,6 @@ public function prepare_items() {
 		$_comments = get_comments( $args );
 
 		if ( is_array( $_comments ) ) {
-			update_comment_cache( $_comments );
-
```
Comment (peterwilsoncc) [inline src/wp-admin/includes/class-wp-comments-list-table.php:169]: This change doesn't seem directly related to this PR, am I missing something? 

I get that they should be cached by the `get_comments()` call.

File: src/wp-includes/comment.php (hunk 9 above)
Comment (peterwilsoncc) [reply src/wp-includes/comment.php:3368]: > We could do that here. If you are a developer that is `update_comment_cache`, it is likely you know what you are doing,I don't want to break that for anyone.

Fair enough, thanks.

File: tests/phpunit/tests/comment/metaCache.php [hunk 16]
```diff
@@ -77,8 +150,6 @@ public function test_update_comment_meta_cache_true() {
 	 * @covers ::update_comment_meta
 	 */
 	public function test_update_comment_meta_cache_false() {
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/comment/metaCache.php]: For cleanliness of commit history, I think the changes below switching from wpdb to get_num_queries would be better done on the generic _unit test improvements for 6.3_ ticket, [WP#57841](https://core.trac.wordpress.org/ticket/57841).

**Review by peterwilsoncc — COMMENTED** [review]
Added a few questions inline but this generally looks pretty good. Thanks.

File: src/wp-admin/includes/class-wp-comments-list-table.php (hunk 15 above)
Comment (spacedmonkey) [reply src/wp-admin/includes/class-wp-comments-list-table.php:169]: It is related. `update_comment_cache` calls `update_meta_cache`, which primes comment meta. This is not needed, as get_comment / WP_Comment_Query does this all for you. This line is a hang over from before `WP_Comment_Query` and is safe to remove now. I reviewed all places that used get_comment / WP_Comment_Query, which is how I found this.

File: tests/phpunit/tests/comment/metaCache.php (hunk 16 above)
Comment (spacedmonkey) [reply tests/phpunit/tests/comment/metaCache.php]: Noted. I will commit this before I commit the rest of the code.

File: src/wp-includes/comment-template.php (hunk 14 above)
Comment (spacedmonkey) [reply src/wp-includes/comment-template.php:2383]: @peterwilsoncc Should we deprecate this in this commit or in another ticket?

`wp_queue_posts_for_term_meta_lazyload` is a different thing, that is primes term meta from a array of posts.

File: tests/phpunit/tests/comment/metaCache.php (hunk 16 above)
Comment (spacedmonkey) [reply tests/phpunit/tests/comment/metaCache.php]: @peterwilsoncc Created a PR for use of `get_num_queries`. https://github.com/WordPress/wordpress-develop/pull/4438

File: src/wp-includes/comment-template.php (hunk 14 above)
Comment (peterwilsoncc) [reply src/wp-includes/comment-template.php:2383]: > Should we deprecate this in this commit or in another ticket?

I'm happy either way.

**Review by peterwilsoncc — APPROVED** [review]
LGTM for commit following the `get_num_queries()` PR.

**Review by felixarntz — APPROVED** [review]
LGTM, thanks @spacedmonkey.

If we commit this as is, it would be great if you could follow up with a ticket to deprecate `wp_queue_comments_for_comment_meta_lazyload()`.

**Comment by spacedmonkey** [comment]
Committed. https://github.com/WordPress/wordpress-develop/commit/3114eda235d6740f82fc9f2b4bdf0735e43560ed
